CONTRIBUTING.md

```diff
@@ -0,0 +1,435 @@
+  a lot of conditional statements in the internal implementation. Finally,
+  it is difficult to assure that all combinations are valid or specify which ones
+  are valid and desired.
+
+- Do not make assumptions about the context in which the component will be used.
+
+  ```qml
+  Item {
+    anchors.fill: parent // Bad, will lead to warnings when used in a Layout
+    // ...
+  }
+  ```
+
+  In most cases, it is a good idea to use `Control` (or more specialized component)
+  as the base component.
+
+  ```qml
+  Control {
+    contentItem: ColumnLayout {
+      // ...
+    }
+  }
+  ```
+
+- Action signals (e.g. click on delegate) should not provide metadata from the
+  model. Only `index` or unique `key` should be an argument. The necessary
+  metadata should be fetched on the signal receiver side.
+
+  ```qml
+  signal collectibleClicked(int chainId, string contractAddress, 
+                            string tokenId, string uid, int tokenType) // BAD
+  signal collectibleClicked(var collectible)                           // BAD
+
+  signal collectibleClicked(string key)                                // OK
+  ```
+  
+  Otherwise, the component unnecessarily bypasses some values from the model.
+  Callers may also want to access roles which are not used for displaying.
+  It leads to creating unnecessary requirements for the input model.
+
+- use consistently `key` as a unique identifier of the model, even if there are
+  other roles with unique values.
+  
+  On the UI side, `key` should be used only for identification, with no other
+  assumptions and usage. Thanks to that, content of that role can be freely
+  changed on the backend with no implications to the UI side. Even if `address`
+  is unique, a separate `key` role should be used (providing the same content as
+  `address` role).
+  
+  Other good name for that role - `id` - does not fit well in qml environment as
+  it is reserved keyword. Defining such role with `ListElement` in tests or
+  Storybook would be not possible.
+  
+- `model.rowCount()` should not be used in bindings.
+  
+  This call provides count only once when it is called, and the expression will not
+  be re-evaluated when count is changed. Solution is to use the attached property
+  `ModelCount` (`model.ModelCount.count`) context property instead. However,
+  in signal handling, using `rowCount()` is fully correct and preferred over `ModelCount`
+  attached property (because it does not create any additional attached object unnecesarily).
+
+- `model.count` should not be used on models taken from outside as a dependency.
+  
+  `count` property is not a part of the `QAbstractItemModel` interface. Proxy
+  models and backend models may not have that property defined.
+
+- Objects holding the whole model's row should not be used in bindings.
+
+  ```qml
+  readonly property var selectedAccount: ModelUtils.get(store.accounts, 0) // BAD
```

Comment:
I guess we decided to use ModelEntry instead of this right? @micieslak

Reply:
Yes, you are right, it would be nice to mention about `ModelEntry` here. I will add along with other amendments.